### ISSUE TITLE
Follow-up on LightAttention adaptation in ThermoMPNN/ThermoMPNN-D

### ISSUE THREAD

**Issue by DRanger666** [Kuhlman-Lab/ThermoMPNN#33]
Hey @hdieckhaus,

Thank you for your detailed and useful reply in #32; it raised some interesting points. I am opening this as a separate issue because the follow-up questions are substantive enough that they might be easier to discuss and find separately, while still continuing directly from #32. Based on your reply there, I looked through this repo and the ThermoMPNN-D repo, and noticed a few things. Your answers would help me get a solid grasp of ThermoMPNN and the double-mutant version, ThermoMPNN-D.

First, in your reply to #32, you said that the kernel_size=9 LA convolutions meant that every convolution received four padded zeroes on each side, "effectively adding a ton of noise (and a lot of parameters)", and that this noise might have helped the model avoid overfitting. However, based on my understanding of the `LightAttention` implementation in `transfer_model.py` ([lines 129-135](https://github.com/Kuhlman-Lab/ThermoMPNN/blob/2b04fd370e399911b1fa5848112cc9013f084110/transfer_model.py#L129-L135)), and how PyTorch `Conv1d` applies kernels along input `dim=2` (`dim=-1`), I don't see that happening. I am saying that because the Conv1d weight tensor has shape `[384, 384, 9]`, and `weight[:, :, 0:4]` and `weight[:, :, 5:9]` are being multiplied by 0 (padded positions) for every data point, which is preventing those parameters from having any impact on the conv layer outputs, and as a result, those extra parameters, although redundant, will not impact active model capacity, and therefore will not have any effect on overfitting, and in principle, should not receive meaningful task-loss gradients. As a matter of fact, the convolutional layers with kernel_size=9 in this context are operationally equivalent to convolutional layers with kernel_size=1, and also equivalent to a 384-to-384 linear layer. This seems coherent with your ThermoMPNN-D observation that kernel_size=1 ([single-mutation path, line 86](https://github.com/Kuhlman-Lab/ThermoMPNN-D/blob/df9a75aaddb674a7c4c193005031fc0536d325fb/thermompnn/model/v2_model.py#L86)) and a Linear layer ([Siamese/double-mutant path, lines 375-382](https://github.com/Kuhlman-Lab/ThermoMPNN-D/blob/df9a75aaddb674a7c4c193005031fc0536d325fb/thermompnn/model/v2_model.py#L375-L382)) resulted in very similar performance to kernel_size=9. **Question:** does my interpretation match the ThermoMPNN/ThermoMPNN-D behavior you observed, or am I missing a mechanism by which those inactive kernel positions could still affect overfitting?

Second, I had analyzed in #32 that the `attention_convolution` -> `Softmax(dim=-1)` sub-block outputs tensors with all 1s due to Softmax application along singleton dimension (irrespective of kernel_size), and you confirmed that understanding in your reply there. This interpretation is applicable to both ThermoMPNN ([LA input construction, lines 103-108](https://github.com/Kuhlman-Lab/ThermoMPNN/blob/2b04fd370e399911b1fa5848112cc9013f084110/transfer_model.py#L103-L108); [`Softmax(dim=-1)` and multiplication, lines 137-154](https://github.com/Kuhlman-Lab/ThermoMPNN/blob/2b04fd370e399911b1fa5848112cc9013f084110/transfer_model.py#L137-L154)) and ThermoMPNN-D ([single-mutation LA input, lines 320-321](https://github.com/Kuhlman-Lab/ThermoMPNN-D/blob/df9a75aaddb674a7c4c193005031fc0536d325fb/thermompnn/model/v2_model.py#L320-L321); [`Softmax(dim=-1)` and multiplication, lines 401-414](https://github.com/Kuhlman-Lab/ThermoMPNN-D/blob/df9a75aaddb674a7c4c193005031fc0536d325fb/thermompnn/model/modules.py#L401-L414)). Since these all 1 `attention` tensors are applied only in the operation

```python
o * self.softmax(attention)
```

in both ThermoMPNN ([line 154](https://github.com/Kuhlman-Lab/ThermoMPNN/blob/2b04fd370e399911b1fa5848112cc9013f084110/transfer_model.py#L154)) and ThermoMPNN-D ([line 414](https://github.com/Kuhlman-Lab/ThermoMPNN-D/blob/df9a75aaddb674a7c4c193005031fc0536d325fb/thermompnn/model/modules.py#L414)), all the parameters of the `attention_convolution` layers have no meaningful impact during forward pass, and in principle, should not receive meaningful task-loss-gradients during backward pass. Therefore, the `attention_convolution` layer does not increase active model capacity. If you observed that this LA block implementation helped avoid overfitting (I am inferring that from your reply), the mechanically consistent explanation seems to be **reduced effective model capacity from inactive LA sub-blocks**. **Question:** does this interpretation align with your observations and understanding around the LA block, particularly its `attention_convolution` -> `Softmax(dim=-1)` sub-block? 

Third, I noticed the citation to Hannes Stark's protein-localization repository ([`LightAttention`, lines 5-48](https://github.com/HannesStark/protein-localization/blob/7b0be1e64a91db8ad1a8feae994a4d09aa9d7b1b/models/light_attention.py#L5-L48)) from your code ([ThermoMPNN, lines 124-127](https://github.com/Kuhlman-Lab/ThermoMPNN/blob/2b04fd370e399911b1fa5848112cc9013f084110/transfer_model.py#L124-L127); [ThermoMPNN-D, lines 389-392](https://github.com/Kuhlman-Lab/ThermoMPNN-D/blob/df9a75aaddb674a7c4c193005031fc0536d325fb/thermompnn/model/modules.py#L389-L392)), and looked through the cited repo. A close inspection tells me that Hannes Stark's input to the LA block is a `[B, C, L]` tensor, where the last dimension indexes `L` residue positions and each position has a `C`-channel representation. By contrast, in ThermoMPNN and ThermoMPNN-D, the last dimension is singleton because one `C`-channel mutation-site representation is passed to the LA block. This discrepancy is the natural outcome of task-specific model-internal representation shape, content, and organization. However, based on my understanding, this discrepancy creates an architectural compatibility issue that requires deliberate reconciliation when the LA block is adapted to process ThermoMPNN's `[B, C, 1]` tensor. The primary reconciliation point is that Hannes Stark's protein-localization problem formulation requires compressing residue-level representation vectors into an instance-level representation, and in that setting, the `attention_convolution` -> `Softmax(dim=-1)` sub-block supplies learned per-channel residue-position weights that are utilized for computing a per-channel weighted average, whereas ThermoMPNN's ProteinMPNN-derived mutation-site representation-to-AA-indexed score-vector mapping sub-problem does not **require** an equivalent step because it already starts with a single mutation-site representation vector. A secondary reconciliation point is that convolving along the residue-position dimension of a `[B, C, L]` tensor with `kernel_size > 1` is coherent with the sequence-neighborhood locality assumption behind a `Conv1d` layer. If the intended ThermoMPNN operation was instead to apply convolution across the 384 embedding coordinates of a `[B, 384, 1]` tensor (roughly speaking, this could be operationalized by passing a second- and third-axis-swapped view of the `[B, 384, 1]` tensor through a `torch.nn.Conv1d` layer configured for that swapped layout), that would require a separate assumption that neighboring embedding coordinates form meaningful local neighborhoods, which is not obviously implied by the ProteinMPNN-derived mutation-site representation. Taken together, these points make your suggestion in #32 that one could use "a bigger MLP or a proper attention layer, rather than this weird LightAttention setup" look like a more coherent and relatively safer alternate implementation direction for ThermoMPNN/ThermoMPNN-D, rather than trying to preserve the original LightAttention architectural components in a setting where the input is, by design, a single ProteinMPNN-derived mutation-site representation with a singleton residue-position axis. **Question:** does this framing match how you now view the adaptation of the original LightAttention block from Hannes Stark's residue-sequence setting to ThermoMPNN's mutation-site representation setting, or was there a specific role you still intended the original LightAttention architectural components to play here?

**Comment by hdieckhaus** [comment]
Hello again,

First, you're correct that the padded parameters are getting zeroed out, though they're still technically affected by gradient updates. This probably doesn't change training behavior much, which explains why the updated kernel_size=1 version is pretty much the same.

Second, based on the fact the attention convolution is pretty much cancelled out, I think you're probably right that the LA block does better than more complicated solutions due to reduced model capacity. I think it's just very easy to overfit on this data if you're not careful.

Third, yes, I think the LightAttention doesn't make much sense when used this way, and an MLP or MultiHeadAttention layer operating on the embedding dimension is a better approach. It may still be useful if applied before the mutation extraction step, when the sequence position is relevant, but we didn't explore that.
